types/ember-data/index.d.ts

```diff
@@ -119,11 +119,13 @@ export namespace DS {
     ): string;
     const VERSION: string;
 
-    interface AttrOptions<T extends Transform | string | number | boolean | Date = any> {
-        defaultValue?: T extends Transform ? (() => ReturnType<T['deserialize']>) : string | number | boolean | (() => T);
+    interface AttrOptions<T> {
+        defaultValue?: T | (() => T);
```

Comment:
We can keep requiring a function for non-primitives with:
```suggestion
        defaultValue?: T extends Exclude<object, null> ? (() => T) : T | (() => T);
```